content/blog/2019-11-08-unified-theory-gc/index.md

```diff
@@ -0,0 +1,245 @@
+of garbage collector, we can see striking similarities between them:
+
+<img src="tracing-wl.png" style="width:100%"> | <img src="rc-wl.png" style="width:100%"> 
+:-------------------------------------:|:--------------------------:
+ Graph traversal algorithm for tracing | Work-list algorithm for RC
+
+The only differences between the two are that tracing increments the reference
+counts of objects on the worklist whereas RC decrements them, and that tracing
+checks if $\rho(w) = 1$ for when to add the objects that $w$ points to whereas
+RC checks if $\rho(w) = 0$. The two garbage collection strategies also differ in
+how $W$ is initialized. For tracing, $W$ initially contains all of the roots of
+the graph, $R$. For reference counting, $W$ starts with all of the objects that
+have had a pointer to them erased since the last time the algorithm ran. By
+viewing the two strategies through this formulation, we can see that they are
+opposites of each other in many ways:
+
+<img src="diff-v2.png" style="width:100%">
+
+As discussed in the mathematical model part, reference counting requires an extra pass to collect cycles. This is generally be done by two strategies: backup tracing, which traces (part of) the heap occasionally, and trial deletion, which attempts to decrease  $\rho(w)$ by trial-and-error guided by heuristics. However, notice that there is also counterpart for that in tracing: sweeping phase. In addition, while tracing converges to the fixed point value stating from the lower bound 0, reference counting starts from the upper bound, which is all incoming pointers that existed historically. 
+
+
+## Hybrids
+
+The authors further show that all realistic garbage collectors are in fact hybrids of tracing and reference counting. In general, we can categorize collectors to unified heap collectors, split heap collectors, and multi-heap collectors. Then different garbage collectors can be seen as performing tracing or reference counting when tracking references within each region and across regions.
+
+### Unified Heap collectors: Deferred Reference Counting & Partial Tracing
+
+<img src="deferred.png" alt="Snow" style="width:100%"> |  <img src="partial.png" alt="Snow" style="width:100%">
+:-------------------------:|:-------------------------:
+Deferred Reference Counting |  Partial Tracing
+
+Rather than doing reference counting completely, Deferred Reference Counting defers updating the reference count of objects pointed directly by roots until batch processing. This is based on the observation that pointers from roots are likely to change very often as they are directly used in the program. Notice that we can view this as tracing from roots to their targets and reference counting for the intra-heap pointers: All the assignments that lead to intra-heap pointers change would be tracked by reference counting as normal. When we suspend the program, we trace the roots for one level, which compensates for the delay.
```

Comment:
```suggestion
Rather than doing reference counting completely, Deferred Reference Counting defers updating the reference count of objects pointed directly by roots until batch processing. This is based on the observation that pointers from roots are likely to change very often as they are directly used in the program. Notice that we can view this as tracing from roots to their targets and reference counting for the intra-heap pointers: All the assignments that lead to intra-heap pointer changes would be tracked by reference counting as normal. When we suspend the program, we trace the roots for one level, which compensates for the delay.
```